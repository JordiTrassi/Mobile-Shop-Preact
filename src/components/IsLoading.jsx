import { CircularProgress, Typography } from '@mui/material';
import { Box } from '@mui/material';


export const IsLoading = () => {
  
  return (
    <Box
      sx={{
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        mt: 20,
      }}>
      <Typography sx={{fontSize: 30}}>Is Loading ...</Typography>
      <CircularProgress sx={{mt: 3}} />
    </Box>
  );
}