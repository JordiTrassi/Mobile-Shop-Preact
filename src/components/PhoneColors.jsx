import { FormControl, FormControlLabel, FormLabel, Radio, RadioGroup } from '@mui/material';
import { useSelector } from 'react-redux';


export const PhoneColors = () => {
    const { colors } = useSelector(state => state.phoneList.selectedPhone);
    console.log(colors);

    return (
      
    <FormControl sx={{p: 2}}>
        <FormLabel id="demo-controlled-radio-buttons-group" sx={{ color: 'white', fontSize: 20}}>Select Color:</FormLabel>
        <RadioGroup
            // aria-labelledby="demo-controlled-radio-buttons-group"
            name="controlled-radio-buttons-group"
            // value={value}
            // onChange={handleChange}
          >
            {                   
                colors.map( color => (
                    <FormControlLabel value={color} control={<Radio />} label={color} sx={{ml: 2}} />    
                    )) 
            }

        </RadioGroup>
    </FormControl>
  )
}
