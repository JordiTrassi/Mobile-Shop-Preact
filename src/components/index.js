export * from './Header';
export * from './IsLoading';
export * from './PhoneCard';
export * from './PhoneColors';
export * from './PhoneInternalMemory';
export * from './ShopTimer';
