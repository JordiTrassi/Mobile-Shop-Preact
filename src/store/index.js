export * from './phoneListSlice';
export * from './store';
export * from './thunks';