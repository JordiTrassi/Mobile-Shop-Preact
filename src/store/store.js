import { configureStore } from '@reduxjs/toolkit';
import { phoneListSlice } from './phoneListSlice';

export const store = configureStore({
    reducer: {
        phoneList: phoneListSlice.reducer,
    },
});