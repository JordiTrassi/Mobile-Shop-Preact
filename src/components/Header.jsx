import { useState, useEffect } from 'preact/hooks';
import { useDispatch } from 'react-redux';
import { startLoadingPhones, getPhones } from '../store';
import { verifyInputValue } from '../helpers/verifyInputValue';

import { AppBar, Badge, Box, IconButton, InputBase, Toolbar, Tooltip, Typography, } from '@mui/material';
import { styled, alpha } from '@mui/material/styles';

import RingVolumeIcon from '@mui/icons-material/RingVolume';
import ShoppingCartIcon from '@mui/icons-material/ShoppingCart';
import SearchIcon from '@mui/icons-material/Search';



const Search = styled('div')(({ theme }) => ({
  position: 'relative',
  borderRadius: theme.shape.borderRadius,
  backgroundColor: alpha(theme.palette.common.white, 0.15),
  '&:hover': {
    backgroundColor: alpha(theme.palette.common.white, 0.25),
  },
  marginRight: theme.spacing(2),
  marginLeft: 0,
  width: '100%',
  [theme.breakpoints.up('sm')]: {
    marginLeft: theme.spacing(3),
    width: 'auto',
  },
}));

const SearchIconWrapper = styled('div')(({ theme }) => ({
  padding: theme.spacing(0, 2),
  height: '100%',
  position: 'absolute',
  pointerEvents: 'none',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
}));

const StyledInputBase = styled(InputBase)(({ theme }) => ({
  color: 'inherit',
  '& .MuiInputBase-input': {
    padding: theme.spacing(1, 1, 1, 0),
    // vertical padding + font size from searchIcon
    paddingLeft: `calc(1em + ${theme.spacing(4)})`,
    transition: theme.transitions.create('width'),
    width: '100%',
    [theme.breakpoints.up('md')]: {
      width: '20ch',
    },
  },
}));

export const Header = () => {

  const dispatch = useDispatch();
  const [inputValue, setInputValue] = useState('');
  
  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onSubmit = async () => {
    const verifiedInputValue = await verifyInputValue(inputValue);
    dispatch(startLoadingPhones({ verifiedInputValue }));
    dispatch(getPhones({ verifiedInputValue }));
  };

  useEffect(() => {
    const listener = event => {
      if (event.code === "Enter" || event.code === "NumpadEnter") {
        event.preventDefault();
        onSubmit();
      }
    };
    document.addEventListener("keydown", listener);
    return () => {
      document.removeEventListener("keydown", listener);
    };
  }, [inputValue]);

  
  return (

    <Box sx={{ flexGrow: 1}}>
      <AppBar position="static" sx={{ borderRadius: 3, backgroundColor: '#4D4D4D' }}>
        <Toolbar>
          <Tooltip
            title="home"
            arrow
            placement="bottom"
          >
            <IconButton
              size="large"
              edge="start"
              color="inherit"
              aria-label="open drawer"
              sx={{ mr: 1 }}
              href="/"
            >
              <RingVolumeIcon />
            </IconButton>
          </Tooltip>    
          <Typography
            variant="h5"
            noWrap
            component="div"
            sx={{ display: { xs: 'none', sm: 'block' } }}
          >
            Mobile Shop
          </Typography>
          <Box sx={{ flexGrow: 1 }} />
          <Search>
            <SearchIconWrapper>
              <SearchIcon />
            </SearchIconWrapper>
            <StyledInputBase
              placeholder="Search mobile..."
              value={inputValue}
              onSubmit={onSubmit}
              onChange={onInputChange}
            />
          </Search>
          <Box sx={{ display: { xs: 'flex', md: 'flex' } }}>
            <Tooltip
            title="Your Cart"
            arrow
            placement="bottom"
            >
              <IconButton
                size="large"
                color="inherit"
              >
                <Badge badgeContent={1} color="error">
                  <ShoppingCartIcon />
                </Badge>
              </IconButton> 
            </Tooltip>
          </Box> 
        </Toolbar>
      </AppBar>
      
    </Box>
  );
}
