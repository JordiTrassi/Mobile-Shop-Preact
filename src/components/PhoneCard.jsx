import { h } from 'preact';
import { useDispatch, useSelector } from 'react-redux';
import { Box, Card, CardContent, CardMedia, IconButton, Tooltip, Typography } from '@mui/material';
import { AddCircle } from '@mui/icons-material';
import { startLoadingSelectedPhone } from '../store/phoneListSlice';
import { getSelectedPhone } from '../store/thunks';
import { useNavigate } from 'react-router-dom';
import { IsLoading } from './IsLoading';


export const PhoneCard = ({
    brand,
    imgUrl,
    model,
    price,
    id,
}) => {

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const { isLoading } = useSelector(state => state.phoneList);


  const onSelectedPhone = () => {

    dispatch(startLoadingSelectedPhone(id));

    (isLoading)
      ? <IsLoading />
      : setTimeout(() => { navigate(`/product/${id}`) }, 150);
    
    dispatch(getSelectedPhone(id));
    
  };
    
  return (
      
      <Card
        className='animate__animated animate__fadeIn'
        sx={{
          display: 'flex',
          mb: '20px',
          justifyContent: 'space-between',
          boxShadow: 6,
          borderRadius: 3,
        }}
      >
      <Box sx={{ display: 'flex', flexDirection: 'column' }} >
        <CardContent sx={{ flex: '1 0 auto' }}>
          <Typography component="div" variant="h4" data-testid="test-collectionName">
            {brand}
          </Typography>
          <Typography component="div" variant="subtitle1" color="text.secondary">
            {model}
          </Typography>
          <Typography component="div" variant="h5" color="text.secondary" sx={{p: 2, ml: 2}}>
            Price: {price} â‚¬
          </Typography>
        </CardContent>
              
        <Box sx={{ display: 'flex', alignItems: 'center', pl: 1, pb: 1 }}>
          <Tooltip
            title="more info"
            placement="right"
            arrow
          >
            <IconButton
              aria-label="btn-detail"
              onClick={onSelectedPhone}
            >
              <AddCircle fontSize= 'large'/> 
            </IconButton>              
          </Tooltip>               
        </Box>
      </Box>
        <CardMedia
          component="img"
          image={imgUrl}
          alt={model}
          sx={{ height: 200, width: 'auto', p: 2 }}     
        />
    </Card>
  );
}