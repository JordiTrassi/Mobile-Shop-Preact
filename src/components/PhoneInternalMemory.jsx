import { useEffect, useState } from 'preact/hooks';
import { useDispatch, useSelector } from 'react-redux';
import { addSelectedPhoneMemory } from '../store/phoneListSlice';
import { FormControl, FormControlLabel, FormLabel, Radio, RadioGroup } from '@mui/material';


export const PhoneInternalMemory = () => {

  const dispatch = useDispatch();
  const { internalMemory } = useSelector(state => state.phoneList.selectedPhone);
  const [memory, setMemory] = useState(internalMemory[0]);
  
  const onSelectedMemory = ({target}) => {
        setMemory(target.value);
        dispatch(addSelectedPhoneMemory(target.value))
  }

  useEffect(() => {
    if (internalMemory.length === 1) {
        dispatch(addSelectedPhoneMemory(memory));
    }
  },[])


  return (
      
    <FormControl sx={{p: 2}}>
        <FormLabel id="demo-controlled-radio-buttons-group" sx={{ color: 'white', fontSize: 20}}>Select Memory:</FormLabel>
        <RadioGroup
            aria-labelledby="demo-controlled-radio-buttons-group"
            name="controlled-radio-buttons-group"
            value={memory}
            onChange={onSelectedMemory}
          >
            {                   
                internalMemory.map( memory => (
                    <FormControlLabel value={memory} control={<Radio />} label={memory} sx={{ml: 2}} />    
                    )) 
            }

        </RadioGroup>
    </FormControl>
  )
}