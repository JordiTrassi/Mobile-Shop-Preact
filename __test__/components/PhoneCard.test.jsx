import { fireEvent, render, screen } from '@testing-library/preact';
import { h } from 'preact';
import { Provider } from 'react-redux';
import { PhoneCard } from '../../src/components/PhoneCard';
import { store } from '../../src/store/store'
import { BrowserRouter } from 'react-router-dom';

const props = {
    brand: 'Samsung',
    imgUrl: 'phone_test_img.jpg',
    model: 'Phone Model Test',
    price: 150,
    id: 'TestId12345',
};

describe('Pruebas en el componente <PhoneCard />', () => {

    beforeEach(() => jest.clearAllMocks());
    
    test('Debe de hacer match con el snapshot con los valores enviados por props', () => {
        
        const { container } = render(
            <BrowserRouter>
                <Provider store={store}>
                    <PhoneCard {...props} />
                </Provider>
            </BrowserRouter> 
        );
        expect(container).toMatchSnapshot();
    });

    test('El iconButton de Material UI debe llamar a onSelectedPhone()', () => {

        // Pasamos como parametro el nombre de la funci√≥n que queremos llamar, 
        // ya que el fireEvent no funciona bien con los componentes de Material UI.  

        const onSelectedPhone = jest.fn();
        const anonymous = jest.fn();

        const configBtnValues = {"l":  {
            "clickfalse": onSelectedPhone,
            "contextmenufalse": anonymous,
            "dragleavefalse": anonymous,
            "focusinfalse": anonymous,
            "focusoutfalse": anonymous,
            "keydownfalse": anonymous,
            "keyupfalse": anonymous,
            "mousedownfalse": anonymous,
            "mouseleavefalse": anonymous,
            "mouseoverfalse": anonymous,
            "mouseupfalse": anonymous,
            "touchendfalse": anonymous,
            "touchmovefalse": anonymous,
            "touchstartfalse": anonymous,
            },
        };
        
        render(
            <BrowserRouter>
                <Provider store={store}>
                    <PhoneCard {...props} />
                </Provider>
            </BrowserRouter>
        );

        const btnDetail = screen.getByRole('button', { name: 'btn-detail' });
        
        // console.log(btnDetail.l.clickfalse);
        // fireEvent.click(btnDetail);

        // expect(onSelectedPhone).toHaveBeenCalled();
       
        expect(btnDetail).toMatchObject(JSON.parse(JSON.stringify(configBtnValues)));  
        
    });
    
     
});
