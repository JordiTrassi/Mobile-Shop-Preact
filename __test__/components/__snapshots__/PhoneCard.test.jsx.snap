// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <PhoneCard /> Debe de hacer match con el snapshot con los valores enviados por props 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root animate__animated animate__fadeIn css-ame1d1-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-j7qwjs"
    >
      <div
        class="MuiCardContent-root css-1qaznj3-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          data-testid="test-collectionName"
        >
          Samsung
        </div>
        <div
          class="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
        >
          Phone Model Test
        </div>
        <div
          class="MuiTypography-root MuiTypography-h5 css-9hy38x-MuiTypography-root"
        >
          Price: 
          150
           â‚¬
        </div>
      </div>
      <div
        class="MuiBox-root css-1ndgitk"
      >
        <button
          aria-label="btn-detail"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <img
      alt="Phone Model Test"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-16z2ejl-MuiCardMedia-root"
      src="phone_test_img.jpg"
    />
  </div>
</div>
`;
